src/types/web-resource.ts

```diff
@@ -0,0 +1,108 @@
+import * as TypeUtils from '../type-utils';
+import { storageRegistry } from '../storage-adapters/index';
+
+/**
+ * This is how WebResources are stored on the DB
+ */
+export type WebResourceRef = {
+	filename: string;
+	href: string;
+	contentType?: string;
+	contentDisposition?: string;
+	size?: number;
+};
+
+/**
+ * This is how WebResources are defined on creation
+ */
+export type WebResourceInput = {
+	filename: string;
+	data: Buffer;
+	contentType?: string;
+	contentDisposition?: string;
+	size?: number;
+	storage: string;
+};
+
+export const types = {
+	postgres: 'JSONB',
+	mysql: 'JSON',
+	websql: 'TEXT',
+	odata: {
+		name: 'Self.WebResource',
+		complexType: `\
+<ComplexType Name="WebResource">
+	<Property Name="filename" Nullable="false" Type="Edm.String"/>\
+	<Property Name="href" Nullable="false" Type="Edm.String"/>\
+	<Property Name="contentType" Nullable="true" Type="Edm.String"/>\
+	<Property Name="contentDisposition" Nullable="true" Type="Edm.String"/>\
+	<Property Name="size" Nullable="true" Type="Edm.Int64"/>\
+</ComplexType>`,
+	},
+};
+
+export const fetchProcessing = (data: any) => {
+	let refData: WebResourceRef;
+	if (data === null) {
+		return data;
+	}
+	if (typeof data === 'string') {
+		refData = JSON.parse(data);
+	} else if (typeof data === 'object') {
+		refData = data;
+	} else {
+		throw new Error(
+			'Unable to parse data for a WebResource type. Data is: ' + data,
+		);
+	}
+
+	return {
+		filename: refData.filename,
+		href: refData.href,
+		contentType: refData.contentType,
+		contentDisposition: refData.contentDisposition,
+		size: refData.size,
+	};
+};
+
+export const validate = TypeUtils.validate.checkRequired(
```

Comment:
@ramirogm We would need a place to transform `multer` file obj to `webresource` type. I took a stab with `validate` function doing that job: 

```ts
export const validate = TypeUtils.validate.checkRequired(
	async (value: WebResourceSource) => {
		/*
		 * Multer file representation:
			[
				{
					"fieldname": "logo",
					"originalname": "logo.png",
					"encoding": "7bit",
					"mimetype": "image/png",
					"buffer": {
					"type": "Buffer",
					"data": [] // being the Blob
					},
					"size": 0
				}
			]
		*/
		if (!_.isObject(value)) {
			throw new Error('received value is not an object');
		}
		if (!_.has(value, 'filename')) {
			throw new Error('filename is required');
		}
		if (!_.has(value, 'data')) {
			throw new Error('data is required');
		}

		// FIXME call the adapter based on parameters
		const webresource = await adapter.saveFile(
			value.fieldname,
			value.buffer.data,
		);

		const refData: WebResourceRefData = {
			filename: value.originalname,
			contentType: value.mimetype,
			contentDisposition: 'attachment',
			href: webresource.href,
			size: value.size,
		};

		return JSON.stringify(refData);
	},
);
```

@Page- Is there a better place to transform the JSON structure?

Reply:
I currently have that code on a draft PR, see https://github.com/balena-io/pinejs/pull/592/files
Now I need to refactor it to move the logic to a proper place, as discussed on thread https://balena.zulipchat.com/#narrow/stream/345890-loop.2Fbalena-io/topic/file-upload/near/311280527